File: tcpcom/EV3/Java/Examples/TrainClient.java
// TrainClient.java

import ch.aplu.ev3.*;
import ch.aplu.tcpcom.*;

public class TrainClient implements TCPClientListener
{
  private TCPClient client;
  private LegoRobot robot;
  private Gear gear;
  private boolean isWaiting = false;
  private long startTime;

  public TrainClient()
  {
    robot = new LegoRobot();
    gear = new Gear();
    robot.addPart(gear);
    gear.setSpeed(25);
    LightSensor lsBottom = new LightSensor(SensorPort.S3);
    robot.addPart(lsBottom);
    LightSensor lsFront = new LightSensor(SensorPort.S1);
    robot.addPart(lsFront);
    String host = "192.168.0.17";
    int port = 5000;
    client = new TCPClient(host, port);
    client.addTCPClientListener(this);
    robot.drawString("Train Client", 0, 1);
    robot.drawString("Trying to connect..", 0, 2);
    boolean success = client.connect();
    if (!success)
    {  
      robot.drawString("Connection failed", 0, 3);
      while (!robot.isEscapeHit())
        continue;
      robot.exit();
      return;
    }
    startTime = System.nanoTime();
    while (!robot.isEscapeHit())
    {
      if (isWaiting)
        gear.stop();
      int vB = lsBottom.getValue();
      int vF = lsFront.getValue();
      if (vB < 500)
        gear.leftArc(0.1);
      else
        gear.rightArc(0.1);
      if (!isWaiting && vF > 500 && System.nanoTime() - startTime > 5000000000L)
      {
        client.sendMessage("go");
        robot.drawString("Stopped", 0, 2);
        isWaiting = true;
      }
    }
    client.disconnect();
    robot.exit();
  }

  public void onStateChanged(String state, String msg)
  {
    if (state.equals(TCPServer.CONNECTED))
    {  
      robot.playTone(1000, 100);
      robot.drawString("Connected", 0, 2);
    }
    else if (state.equals(TCPClient.DISCONNECTED))
    {  
      robot.playTone(500, 100);
      robot.drawString("Disconnected", 0, 2);
      isWaiting = true;
    }
    else if (state.equals(TCPServer.MESSAGE))
    {
      isWaiting = false;
      startTime = System.nanoTime();
      robot.drawString("Running", 0, 2);
    }
  }

  public static void main(String[] args)
  {
    new TrainClient();
  }
}

File: tcpcom/Pi2Go/TigerJython/Examples/TrainClient.py
# TrainClient.py

from raspibrick import *
from tcpcom import TCPClient
import time

def onStateChanged(state, msg):
    global isWaiting, start
    if state == TCPClient.CONNECTED:
        ledLeft.setColor("green")
    if state == TCPClient.DISCONNECTED:
        ledLeft.setColor("red")
        gear.stop()
    if state == TCPClient.MESSAGE:
        isWaiting = False
        start = time.time()
        display.showText("run")

robot = Robot()
display = Display()
ledLeft = Led(LED_LEFT)
ledLeft.setColor("red")
gear = Gear()
gear.setSpeed(25)
irLeft = InfraredSensor(IR_LINE_LEFT)
irRight = InfraredSensor(IR_LINE_RIGHT)
irCenter = InfraredSensor(IR_CENTER)
r = 0.1
host = "192.168.0.4"
port = 5000
client = TCPClient(host, port, stateChanged = onStateChanged)
rc = client.connect()
isWaiting = False
display.showText("try")
if rc:
    display.showText("run")
    start = time.time()
    while not isEscapeHit():
        if isWaiting:
            continue
        vL = irLeft.getValue()
        vR = irRight.getValue()
        vC = irCenter.getValue()
        if vL == 1 and vR == 0:
            gear.forward()
        elif vL == 0 and vR == 0:   
            gear.leftArc(r)
        elif vL == 1 and vR == 1:
            gear.rightArc(r)
        if vC == 1 and time.time() - start > 5:
            client.sendMessage("go")
            display.showText("stop")
            gear.stop()
            isWaiting = True
    client.disconnect()                
robot.exit()
